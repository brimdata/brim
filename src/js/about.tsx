import React from "react"

import AboutWindow from "./components/AboutWindow"
import lib from "./lib"
import {createRoot} from "react-dom/client"

const container = lib.doc.id("about-root")
const root = createRoot(container!)
root.render(<AboutWindow />)
