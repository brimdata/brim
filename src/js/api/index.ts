import {remote} from "electron"
import path from "path"
import toast from "react-hot-toast"
import {getZealot} from "../flows/getZealot"
import {AppDispatch, State} from "../state/types"
import Current from "../state/Current"
import {CommandRegistry, LoaderRegistry} from "./registries"
import {ToolbarApi} from "./ui-apis"

export default class BrimApi {
  public commands = new CommandRegistry()
  public loaders = new LoaderRegistry()
  public toolbar = new ToolbarApi()
  // public contextMenu = new ContextMenuApi()
  public toast = toast

  /*
  TODO: store is made public only to make the initial plugin architecture
    migration easier, it will eventually be private
   */
  public dispatch: AppDispatch
  public getState: () => State

  constructor() {}

  public setStoreArgs(d: AppDispatch, gs: () => State) {
    this.dispatch = d
    this.getState = gs

    this.toolbar.setStoreArgs(d, gs)
  }

  public getCurrent() {
    const state = this.getState()
    const space = Current.getSpace(state)
    const ws = Current.getWorkspace(state)
    return {
      spaceId: space.id,
      spaceName: space.name,
      workspaceId: ws.id,
      workspaceName: ws.name
    }
  }

  public getZealot() {
    return this.dispatch(getZealot())
  }

  public getTempDir() {
    return remote.app.getPath("temp")
  }

  public getAppConfig() {
    return {
      dataRoot: path.join(remote.app.getPath("userData"), "data"),
      zdepsDirectory: path.join(
        remote.app.getAppPath().replace("app.asar", "app.asar.unpacked"),
        "zdeps"
      )
    }
  }
}
