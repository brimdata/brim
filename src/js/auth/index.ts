import jwtDecode from "jwt-decode"
import url from "url"
import keytar from "keytar"
import os from "os"
import got from "got"
import {shell} from "electron"
import "regenerator-runtime/runtime"

export class Authenticator {
  private accessToken: string
  private refreshToken: string
  private profile: any

  private apiIdentifier = "https://app.brimsecurity.com"
  private redirectUri = "brim:///"

  private keytarServiceSuffix = "-brim-oath"
  private keytarAccount = os.userInfo().username

  constructor(
    private workspaceUrl: string,
    private clientId: string,
    private auth0Domain: string
  ) {}

  getAccessToken(): string {
    return this.accessToken
  }

  // TODO: where will we use profile?
  getProfile(): string {
    return this.profile
  }

  login(): Promise<void> {
    return shell.openExternal(
      "https://" +
        this.auth0Domain +
        "/authorize?" +
        "audience=" +
        this.apiIdentifier +
        "&" +
        "scope=openid profile offline_access&" +
        "response_type=code&" +
        "client_id=" +
        this.clientId +
        "&" +
        "redirect_uri=" +
        this.redirectUri
    )
  }

  private getKeytarService(): string {
    const service = `${this.workspaceUrl}${this.keytarServiceSuffix}`
    // TODO: MASON - remove me
    console.log("keytar service is: ", service)
    return service
  }

  async refreshTokens(): Promise<void> {
    this.refreshToken = await keytar.getPassword(
      this.getKeytarService(),
      this.keytarAccount
    )

    if (this.refreshToken) {
      try {
        const data = await got.post(`https://${this.auth0Domain}/oauth/token`, {
          headers: {
            "content-type": "application/json"
          },
          json: {
            grant_type: "refresh_token",
            client_id: this.clientId,
            refresh_token: this.refreshToken
          },
          responseType: "json"
        })

        const {body} = data
        // @ts-ignore
        this.accessToken = body.access_token
        // @ts-ignore
        this.profile = jwtDecode(body.id_token)
      } catch (error) {
        await this.logout()

        throw error
      }
    } else {
      throw new Error("No available refresh token.")
    }
  }

  async loadTokens(callbackURL): Promise<void> {
    const urlParts = url.parse(callbackURL, true)
    const query = urlParts.query

    const exchangeOptions = {
      grant_type: "authorization_code",
      client_id: this.clientId,
      code: query.code,
      redirect_uri: this.redirectUri
    }

    const fetchUrl = `https://${this.auth0Domain}/oauth/token`

    try {
      const data = await got.post(fetchUrl, {
        headers: {
          "content-type": "application/json"
        },
        json: exchangeOptions,
        responseType: "json"
      })
      const {body} = data

      // TODO: define expected type for body
      // @ts-ignore
      this.accessToken = body.access_token
      // @ts-ignore
      this.profile = jwtDecode(body.id_token)
      // @ts-ignore
      this.refreshToken = body.refresh_token

      if (this.refreshToken) {
        console.log("setting password: ", this.getKeytarService())
        await keytar.setPassword(
          this.getKeytarService(),
          this.keytarAccount,
          this.refreshToken
        )
      }
    } catch (error) {
      await this.logout()

      throw error
    }
  }

  async logout(): Promise<void> {
    await keytar.deletePassword(this.getKeytarService(), this.keytarAccount)
    this.accessToken = null
    this.profile = null
    this.refreshToken = null
  }

  // getLogOutUrl(): string {
  //   return `https://${this.auth0Domain}/v2/logout`
  // }
}
