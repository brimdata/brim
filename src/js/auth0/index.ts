import jwtDecode from "jwt-decode"
import url from "url"
import keytar from "keytar"
import os from "os"
import got from "got"
import {shell} from "electron"
// import "regenerator-runtime/runtime"

const audience = "https://app.brimsecurity.com"
const redirectUri = "brim://auth0/callback"

const keytarServiceSuffix = "brim-oauth"

export const toAccessTokenKey = (id: string) =>
  [id, "AT", keytarServiceSuffix].join("-")
export const toRefreshTokenKey = (id: string) =>
  [id, "RT", keytarServiceSuffix].join("-")

const Auth0 = ()

export class Authenticator {
  private accessToken: string
  private refreshToken: string
  private profile: any

  private apiIdentifier = "https://app.brimsecurity.com"
  private redirectUri = "brim://callback"

  private keytarServiceSuffix = "-brim-oauth"
  private keytarAccount = os.userInfo().username

  constructor(
    private workspaceId: string,
    private clientId: string,
    private auth0Domain: string
  ) {}

  static serializeState(...stateItems: string[]): string {
    return stateItems.join(",")
  }
  static deserializeState(state: string): string[] {
    return state.split(",")
  }

  getAccessToken(): string {
    return this.accessToken
  }

  // TODO: where will we use profile?
  getProfile(): string {
    return this.profile
  }

  login(state: string): Promise<void> {
    const loginUrl =
      this.auth0Domain +
      "/authorize?" +
      "audience=" +
      this.apiIdentifier +
      "&" +
      "scope=openid profile offline_access&" +
      "response_type=code&" +
      "client_id=" +
      this.clientId +
      "&" +
      "redirect_uri=" +
      this.redirectUri +
      "&" +
      "state=" +
      state

    return shell.openExternal(loginUrl)
  }

  private getKeytarService(): string {
    const service = this.workspaceId + this.keytarServiceSuffix
    return service
  }

  async refreshTokens(): Promise<string> {
    this.refreshToken = await keytar.getPassword(
      this.getKeytarService(),
      this.keytarAccount
    )

    if (this.refreshToken) {
      try {
        const data = await got.post(`https://${this.auth0Domain}/oauth/token`, {
          headers: {
            "content-type": "application/json"
          },
          json: {
            grant_type: "refresh_token",
            client_id: this.clientId,
            refresh_token: this.refreshToken
          },
          responseType: "json"
        })

        const {body} = data
        // @ts-ignore
        this.accessToken = body.access_token
        // @ts-ignore
        this.profile = jwtDecode(body.id_token)
      } catch (error) {
        await this.logout()

        throw error
      }
    } else {
      throw new Error("No available refresh token.")
    }

    return this.accessToken
  }

  async loadTokens(callbackURL): Promise<string> {
    const urlParts = url.parse(callbackURL, true)
    const query = urlParts.query

    const exchangeOptions = {
      grant_type: "authorization_code",
      client_id: this.clientId,
      code: query.code,
      redirect_uri: this.redirectUri
    }

    const fetchUrl = `${this.auth0Domain}/oauth/token`

    try {
      const data = await got.post(fetchUrl, {
        headers: {
          "content-type": "application/json"
        },
        json: exchangeOptions,
        responseType: "json"
      })
      const {body} = data

      // TODO: define expected type for body
      // @ts-ignore
      this.accessToken = body.access_token
      // @ts-ignore
      this.profile = jwtDecode(body.id_token)
      // @ts-ignore
      this.refreshToken = body.refresh_token

      if (this.refreshToken) {
        await keytar.setPassword(
          this.getKeytarService(),
          this.keytarAccount,
          this.refreshToken
        )
      }
    } catch (error) {
      await this.logout()

      throw error
    }

    return this.accessToken
  }

  async logout(): Promise<void> {
    await keytar.deletePassword(this.getKeytarService(), this.keytarAccount)
    this.accessToken = null
    this.profile = null
    this.refreshToken = null
  }

  // getLogOutUrl(): string {
  //   return `${this.auth0Domain}/v2/logout`
  // }
}
