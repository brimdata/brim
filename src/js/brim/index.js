/* @flow */
import type {FieldValue} from "../types/records"
import type {ReturnType} from "../types"
import ast from "./ast"
import compoundField from "./compoundField"
import dateTuple from "./dateTuple"
import entries from "./entries"
import field from "./field"
import flatRecordsBuffer from "./flatRecordsBuffer"
import form from "./form"
import interop from "./interop"
import log from "./log"
import program from "./program"
import randomHash from "./randomHash"
import record from "./record"
import relTime from "./relTime"
import search from "./search"
import space from "./space"
import span from "./span"
import syntax from "./syntax"
import tab from "./brimTab"
import table from "./table"
import time from "./time"
import zeekLogInfo from "./zeekLogInfo"

export type $Field = {
  name: string,
  type: string,
  value: FieldValue,
  stringValue: () => string,
  queryableValue: () => string,
  compound: () => boolean,
  toCompound: () => $CompoundField,
  toDate: () => Date,
  display: () => string,
  guessWidth: () => number
}

export type $CompoundField = ReturnType<typeof compoundField>
export type $Search = ReturnType<typeof search>
export type $Log = ReturnType<typeof log>

export type Ts = {
  ns: number,
  sec: number
}

export type Span = [Ts, Ts]

export default {
  table,
  program,
  field,
  compoundField,
  log,
  ast,
  syntax,
  space,
  time,
  relTime,
  span,
  dateTuple,
  flatRecordsBuffer,
  form,
  record,
  interop,
  search,
  randomHash,
  entries,
  tab,
  zeekLogInfo
}
