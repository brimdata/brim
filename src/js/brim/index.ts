import ast from "./ast"
import dateTuple from "./dateTuple"
import entries from "./entries"
import form from "./form"
import interop from "./interop"
import program from "./program"
import randomHash from "./randomHash"
import relTime from "./relTime"
import space from "./space"
import span from "./span"
import syntax from "./syntax"
import tab from "./brimTab"

import time from "./time"
import zeekLogInfo from "./zeekLogInfo"
import connection from "./connection"

export type Ts = {
  ns: number
  sec: number
}

export type BrimSpace = ReturnType<typeof space>
export type BrimConnection = ReturnType<typeof connection>

export type Span = [Ts, Ts]

export default {
  program,
  ast,
  syntax,
  space,
  time,
  relTime,
  span,
  dateTuple,
  form,
  interop,
  randomHash,
  entries,
  tab,
  zeekLogInfo,
  connection
}
