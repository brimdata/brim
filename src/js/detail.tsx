import "regenerator-runtime/runtime"

import {Provider} from "react-redux"
import React from "react"
import ReactDOM from "react-dom"

import AppErrorBoundary from "./components/AppErrorBoundary"
import initDetail from "./initializers/initDetail"
import lib from "./lib"
import LogDetailsWindow from "./components/LogDetailsWindow"
import StartupError from "./components/StartupError"
import {Modals} from "./components/Modals"
import {ThemeProvider} from "styled-components"
import theme from "./style-theme"
import BrimTooltip from "./components/BrimTooltip"

initDetail()
  .then((store) => {
    ReactDOM.render(
      <AppErrorBoundary>
        <div id="modal-dialog-root" />
        <Provider store={store}>
          <ThemeProvider theme={theme}>
            <LogDetailsWindow />
            <Modals />
            <BrimTooltip />
          </ThemeProvider>
        </Provider>
      </AppErrorBoundary>,
      lib.doc.id("app-root")
    )
  })
  .catch((e) => {
    ReactDOM.render(<StartupError error={e} />, lib.doc.id("app-root"))
  })
