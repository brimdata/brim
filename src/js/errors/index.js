/* @flow */
import logsIngest from "./logsIngest"
import pcapIngest from "./pcapIngest"

export default {
  pcapIngest,
  logsIngest,
  spaceDeleted: (name: string) => ({
    type: "SpaceDeletedError",
    message: `The space "${name}" previously on this tab has been deleted.`
  })
}
