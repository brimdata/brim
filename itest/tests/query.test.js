/* @flow */

import {basename} from "path"

import {
  getSearchSpeed,
  getSearchTime,
  logIn,
  newAppInstance,
  resetState,
  searchDisplay,
  setSpan,
  startApp,
  startSearch,
  waitForSearch,
  writeSearch
} from "../lib/app.js"
import {handleError, stdTest} from "../lib/jest.js"

describe("Query tests", () => {
  let app
  let testIdx = 0
  beforeEach(() => {
    app = newAppInstance(basename(__filename), ++testIdx)
    return startApp(app)
  })

  afterEach(async () => {
    if (app && app.isRunning()) {
      await resetState(app)
      return await app.stop()
    }
  })

  stdTest('query "_path=weird | sort ts"', (done) => {
    logIn(app)
      .then(() => writeSearch(app, "_path=weird | sort ts"))
      .then(() => startSearch(app))
      .then(() => waitForSearch(app))
      .then(() => searchDisplay(app))
      .then((results) => {
        expect(results).toMatchSnapshot()
      })
      .then(() => getSearchSpeed(app))
      .then((searchSpeed) => {
        expect(searchSpeed).toBeGreaterThan(0)
        expect(searchSpeed).toBeLessThan(1000)
      })
      .then(() => getSearchTime(app))
      .then((searchTime) => {
        expect(searchTime).toBeGreaterThan(0)
        expect(searchTime).toBeLessThan(5)
        done()
      })
      .catch((err) => {
        handleError(app, err, done)
      })
  })

  stdTest('query "_path=http | count()"', (done) => {
    logIn(app)
      .then(() => writeSearch(app, "_path=http | count()"))
      .then(() => startSearch(app))
      .then(() => waitForSearch(app))
      .then(() => searchDisplay(app))
      .then((results) => {
        expect(results).toMatchSnapshot()
      })
      .then(() => getSearchSpeed(app))
      .then((searchSpeed) => {
        expect(searchSpeed).toBeGreaterThan(0)
        expect(searchSpeed).toBeLessThan(1000)
      })
      .then(() => getSearchTime(app))
      .then((searchTime) => {
        expect(searchTime).toBeGreaterThan(0)
        expect(searchTime).toBeLessThan(5)
        done()
      })
      .catch((err) => {
        handleError(app, err, done)
      })
  })

  stdTest('query "_path=http | count() by id.resp_p | sort count"', (done) => {
    logIn(app)
      .then(() =>
        writeSearch(app, "_path=http | count() by id.resp_p | sort count")
      )
      .then(() => startSearch(app))
      .then(() => waitForSearch(app))
      .then(() => searchDisplay(app))
      .then((results) => {
        expect(results).toMatchSnapshot()
      })
      .then(() => getSearchSpeed(app))
      .then((searchSpeed) => {
        expect(searchSpeed).toBeGreaterThan(0)
        expect(searchSpeed).toBeLessThan(1000)
      })
      .then(() => getSearchTime(app))
      .then((searchTime) => {
        expect(searchTime).toBeGreaterThan(0)
        expect(searchTime).toBeLessThan(5)
        done()
      })
      .catch((err) => {
        handleError(app, err, done)
      })
  })

  stdTest('query "_path=http | every 5m count()"', (done) => {
    logIn(app)
      .then(() => writeSearch(app, "_path=http | every 5m count()"))
      .then(() => startSearch(app))
      .then(() => waitForSearch(app))
      .then(() => searchDisplay(app))
      .then((results) => {
        expect(results).toMatchSnapshot()
      })
      .then(() => getSearchSpeed(app))
      .then((searchSpeed) => {
        expect(searchSpeed).toBeGreaterThan(0)
        expect(searchSpeed).toBeLessThan(1000)
      })
      .then(() => getSearchTime(app))
      .then((searchTime) => {
        expect(searchTime).toBeGreaterThan(0)
        expect(searchTime).toBeLessThan(5)
        done()
      })
      .catch((err) => {
        handleError(app, err, done)
      })
  })

  stdTest('query "* | count()"; switch to whole space', (done) => {
    logIn(app)
      .then(() => writeSearch(app, "* | count()"))
      .then(() => startSearch(app))
      .then(() => waitForSearch(app))
      .then(() => searchDisplay(app))
      .then((results) => {
        expect(results).toMatchSnapshot()
      })
      .then(() => setSpan(app, "Whole Space"))
      .then(() => waitForSearch(app))
      .then(() => searchDisplay(app))
      .then((results) => {
        expect(results).toMatchSnapshot()
      })
      .then(() => getSearchSpeed(app))
      .then((searchSpeed) => {
        expect(searchSpeed).toBeGreaterThan(0)
        expect(searchSpeed).toBeLessThan(1000)
      })
      .then(() => getSearchTime(app))
      .then((searchTime) => {
        expect(searchTime).toBeGreaterThan(0)
        expect(searchTime).toBeLessThan(5)
        done()
      })
      .catch((err) => {
        handleError(app, err, done)
      })
  })

  stdTest(
    'query "_path=x509 | every 1d count() by certificate.version | sort ts, certificate.version"',
    (done) => {
      logIn(app)
        .then(() =>
          writeSearch(
            app,
            "_path=x509 | every 1d count() by certificate.version | sort ts, certificate.version"
          )
        )
        .then(() => setSpan(app, "Whole Space"))
        .then(() => startSearch(app))
        .then(() => waitForSearch(app))
        .then(() => searchDisplay(app))
        .then((results) => {
          expect(results).toMatchSnapshot()
          done()
        })
        .catch((err) => {
          handleError(app, err, done)
        })
    }
  )
})
