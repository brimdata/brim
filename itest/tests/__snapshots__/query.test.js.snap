// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query tests query000: "* | count()" 1`] = `
Array [
  "count",
  "30",
]
`;

exports[`Query tests query001: "* | count() by _path | sort _path" 1`] = `
Array [
  "_path",
  "count",
  "capture_loss",
  "1",
  "conn",
  "19",
  "dns",
  "2",
  "files",
  "4",
  "ssl",
  "1",
  "stats",
  "1",
  "weird",
  "1",
  "x509",
  "1",
]
`;

exports[`Query tests query002: "_path=conn | count()" 1`] = `
Array [
  "count",
  "19",
]
`;

exports[`Query tests query003: "_path=conn | cut ts, id.orig_h, id.orig_p, id.resp_h, id.resp_p, proto | sort ts" 1`] = `
Array [
  "ts",
  "id.orig_h",
  "id.orig_p",
  "id.resp_h",
  "id.resp_p",
  "proto",
  "2020-02-25T16:03:05Z",
  "192.168.1.110",
  "56625",
  "172.217.9.142",
  "443",
  "tcp",
  "2020-02-25T16:03:06Z",
  "192.168.1.110",
  "57591",
  "172.217.9.142",
  "443",
  "tcp",
  "2020-02-25T16:03:07Z",
  "192.168.1.110",
  "55635",
  "18.246.31.137",
  "443",
  "tcp",
  "2020-02-25T16:03:07Z",
  "192.168.1.110",
  "55747",
  "13.52.5.22",
  "443",
  "tcp",
  "2020-02-25T16:03:08Z",
  "192.168.1.110",
  "55344",
  "52.37.243.173",
  "443",
  "tcp",
  "2020-02-25T16:03:08Z",
  "192.168.1.110",
  "55346",
  "52.37.243.173",
  "443",
  "tcp",
  "2020-02-25T16:03:08Z",
  "192.168.1.110",
  "57326",
  "173.194.201.189",
  "443",
  "tcp",
  "2020-02-25T16:03:08Z",
  "192.168.1.110",
  "57487",
  "192.30.253.125",
  "443",
  "tcp",
  "2020-02-25T16:03:08Z",
  "192.168.1.110",
  "55354",
  "52.37.243.173",
  "443",
  "tcp",
  "2020-02-25T16:03:09Z",
  "192.168.1.179",
  "47783",
  "192.168.1.255",
  "15600",
  "udp",
  "2020-02-25T16:03:10Z",
  "192.168.1.110",
  "57332",
  "64.233.179.189",
  "443",
  "tcp",
  "2020-02-25T16:03:10Z",
  "192.168.1.110",
  "55351",
  "18.205.93.211",
  "443",
  "tcp",
  "2020-02-25T16:03:11Z",
  "192.168.1.110",
  "54375",
  "192.168.1.254",
  "53",
  "udp",
  "2020-02-25T16:03:12Z",
  "192.168.1.110",
  "57635",
  "17.125.252.5",
  "443",
  "tcp",
  "2020-02-25T16:03:13Z",
  "192.168.1.110",
  "51848",
  "192.168.1.254",
  "53",
  "udp",
  "2020-02-25T16:03:13Z",
  "192.168.1.110",
  "57640",
  "209.216.230.240",
  "443",
  "tcp",
  "2020-02-25T16:03:15Z",
  "192.168.1.179",
  "51524",
  "192.168.1.255",
  "15600",
  "udp",
  "2020-02-25T16:03:15Z",
  "192.168.1.110",
  "57572",
  "172.217.6.138",
  "443",
  "tcp",
  "2020-02-25T16:03:15Z",
  "192.168.1.110",
  "57540",
  "172.217.1.138",
  "443",
  "tcp",
]
`;

exports[`Query tests query004: "* | every 2s count() | sort ts" 1`] = `
Array [
  "ts",
  "count",
  "2020-02-25T16:03:04Z",
  "2",
  "2020-02-25T16:03:06Z",
  "4",
  "2020-02-25T16:03:08Z",
  "6",
  "2020-02-25T16:03:10Z",
  "4",
  "2020-02-25T16:03:12Z",
  "4",
  "2020-02-25T16:03:14Z",
  "9",
  "2020-02-25T16:03:16Z",
  "1",
]
`;

exports[`Query tests query005: "* | every 2s count() by _path | sort ts, _path" 1`] = `
Array [
  "ts",
  "_path",
  "count",
  "2020-02-25T16:03:04Z",
  "conn",
  "1",
  "2020-02-25T16:03:04Z",
  "stats",
  "1",
  "2020-02-25T16:03:06Z",
  "conn",
  "3",
  "2020-02-25T16:03:06Z",
  "weird",
  "1",
  "2020-02-25T16:03:08Z",
  "conn",
  "6",
  "2020-02-25T16:03:10Z",
  "conn",
  "3",
  "2020-02-25T16:03:10Z",
  "dns",
  "1",
  "2020-02-25T16:03:12Z",
  "conn",
  "3",
  "2020-02-25T16:03:12Z",
  "dns",
  "1",
  "2020-02-25T16:03:14Z",
  "conn",
  "3",
  "2020-02-25T16:03:14Z",
  "files",
  "4",
  "2020-02-25T16:03:14Z",
  "ssl",
  "1",
  "2020-02-25T16:03:14Z",
  "x509",
  "1",
  "2020-02-25T16:03:16Z",
  "capture_loss",
  "1",
]
`;
